tests/handover.cpp

```diff
@@ -0,0 +1,370 @@
+#include "object_schema.hpp"
+#include "property.hpp"
+#include "schema.hpp"
+#include "handover.hpp"
+
+#include <thread>
+#include <future>
+
+using namespace realm;
+
+static TableRef get_table(Realm& realm, const ObjectSchema &object_schema) {
+    StringData name (("class_" + object_schema.name).c_str());
+    return realm.read_group().get_table(name);
+}
+
+static Object create_object(SharedRealm realm, const ObjectSchema &object_schema) {
+    TableRef table = get_table(*realm, object_schema);
+    return Object(std::move(realm), object_schema, Row((*table)[table->add_empty_row()]));
+}
+
+static List get_list(const Object& object, size_t column_ndx) {
+    return List(object.realm(), object.row().get_table()->get_linklist(column_ndx, object.row().get_index()));
+}
+
+static Property nullable(Property p) {
+    p.is_nullable = true;
+    return p;
+}
+
+TEST_CASE("handover") {
+    InMemoryTestFile config;
+    config.cache = false;
+    config.automatic_change_notifications = false;
+
+    SharedRealm r = Realm::get_shared_realm(config);
+
+    static const ObjectSchema string_object ({"string_object", {
+        nullable({"value", PropertyType::String}),
+    }});
+    static const ObjectSchema int_object ({"int_object", {
+        {"value", PropertyType::Int},
+    }});
+    static const ObjectSchema int_array_object ({"int_array_object", {
+        {"value", PropertyType::Array, "int_object"}
+    }});
+    r->update_schema({string_object, int_object, int_array_object});
+
+    SECTION("disallowed during write transactions") {
+        SECTION("export") {
+            r->begin_transaction();
+            REQUIRE_THROWS(r->package_for_handover({}));
+        }
+        SECTION("import") {
+            auto h = r->package_for_handover({});
+            r->begin_transaction();
+            REQUIRE_THROWS(r->accept_handover(std::move(h)));
+        }
+    }
+
+    SECTION("cleanup properly unpins version") {
+        auto durability = config.in_memory ? SharedGroup::durability_MemOnly :
+        SharedGroup::durability_Full;
+        auto history = realm::make_client_history(config.path, config.encryption_key.data());
+        SharedGroup shared_group(*history, durability);
+
+        auto get_current_version = [&]() -> SharedGroup::VersionID {
+            shared_group.begin_read();
+            auto version = shared_group.get_version_of_current_transaction();
+            shared_group.end_read();
+            return version;
+        };
+
+        auto handover_version = get_current_version();
+        auto h = r->package_for_handover({});
+        r->begin_transaction(); r->commit_transaction(); // Advance version
+
+        REQUIRE(get_current_version() != handover_version); // Ensure advanced
+        REQUIRE_NOTHROW(shared_group.begin_read(handover_version)); shared_group.end_read(); // Ensure pinned
+
+        SECTION("destroyed without being imported") {
+            h.~HandoverPackage(); // Destory handover, unpinning version
```

Comment:
This results in the destructor being called twice on the same object, which is UB. Could use a util::Optional<Handover> instead to let you destroy it here in a more natural way.